crypto/threads_pthread.c

```diff
@@ -42,12 +45,563 @@
+    pthread_mutex_t prior_lock;
+
+    /* signal to wake threads waiting on prior_lock */
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data =
+                        (struct rcu_thr_data *)CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    OPENSSL_free(data);
+    CRYPTO_THREAD_set_local(&rcu_thr_key, NULL);
+}
+
+static void ossl_rcu_init(void)
+{
+    CRYPTO_THREAD_init_local(&rcu_thr_key, NULL);
+}
+
+/* Read side acquisition of the current qp */
+static struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK lock)
+{
+    uint64_t qp_idx;
+
+    /* get the current qp index */
+    for (;;) {
+        /*
+         * Notes on use of __ATOMIC_ACQUIRE
+         * We need to ensure the following:
+         * 1) That subsequent operations aren't optimized by hoisting them above
+         * this operation.  Specifically, we don't want the below re-load of
+         * qp_idx to get optimized away
+         * 2) We want to ensure that any updating of reader_idx on the write side
+         * of the lock is flushed from a local cpu cache so that we see any
+         * updates prior to the load.  This is a non-issue on cache coherent
+         * systems like x86, but is relevant on other arches
+         * Note: This applies to the reload below as well
+         */
+        qp_idx = (uint64_t)ATOMIC_LOAD_N(&lock->reader_idx, __ATOMIC_ACQUIRE);
+
+        /*
+         * Notes of use of __ATOMIC_RELEASE
+         * This counter is only read by the write side of the lock, and so we
+         * specify __ATOMIC_RELEASE here to ensure that the write side of the
+         * lock see this during the spin loop read of users, as it waits for the
+         * reader count to approach zero
+         */
+        ATOMIC_ADD_FETCH(&lock->qp_group[qp_idx].users, VAL_READER,
+                         __ATOMIC_RELEASE);
+
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)ATOMIC_LOAD_N(&lock->reader_idx, __ATOMIC_ACQUIRE))
+            break;
+
+        /*
+         * Notes on use of __ATOMIC_RELEASE
+         * As with the add above, we want to ensure that this decrement is
+         * seen by the write side of the lock as soon as it happens to prevent
+         * undue spinning waiting for write side completion
+         */
+        ATOMIC_SUB_FETCH(&lock->qp_group[qp_idx].users, VAL_READER,
+                         __ATOMIC_RELEASE);
+    }
+
+    return (struct rcu_qp *)&lock->qp_group[qp_idx];
+}
+
+void ossl_rcu_read_lock(CRYPTO_RCU_LOCK lock)
+{
+    struct rcu_thr_data *data;
+    int i;
+    int available_qp = -1;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data == NULL) {
+        for (;;) {
```

Comment:
Why this `for (;;)` loop , just to allocate your rcu_thr_data struct ?

Reply:
The desire is to allow ossl_rcu_read_lock to not fail, and so the only option here is to repeatedly try to allocate the threads qp data until it succeeds

Reply:
Hmm, could we use OPENSSL_assert() instead? Or maybe a few loops (perhaps with some sleep() in between) and then assert?

Reply:
I do not think keeping the loop as is is a good idea as it basically converts an OOM into a busy loop.

Or make the function return an error as OOM should be an error return and not causing an abort().

Reply:
the assert approach would probably work best, as I'd like the lock operation to return void if possible

Reply:
corrected to use OPENSSL_assert